appinventor/components/src/com/google/appinventor/components/runtime/ChartDataModel.java

```diff
@@ -0,0 +1,554 @@
+    // Since real-time data comes in fast, the case of
+    // multi-data source input is unhandled since it's
+    // better to avoid it.
+    protected int maximumTimeEntries = 200;
+
+    /**
+     * Enum used to specify the criterion to use for entry filtering/comparing.
+     */
+    public enum EntryCriterion {
+        All, // Return all entries
+        XValue,
+        YValue;
+    }
+
+    /**
+     * Initializes a new ChartDataModel object instance.
+     *
+     * @param data  Chart data instance
+     */
+    protected ChartDataModel(D data) {
+        this.data = data;
+        entries = new ArrayList<Entry>();
+    }
+
+    /**
+     * Returns the size of the tuples that this Data Series
+     * accepts.
+     *
+     * @return  tuple size (integer)
+     */
+    protected abstract int getTupleSize();
+
+    /**
+     * Returns the Data Series of the Data Model.
+     * The method is made synchronized to avoid concurrent
+     * access between threads, which can cause exceptions
+     * if multiple threads try to modify the Dataset object
+     * at the same time.
+     * @return  Data Series object of the Data model
+     */
+    public synchronized T getDataset() {
+        return dataset;
+    }
+
+    public ChartData getData() {
+        return data;
+    }
+
+    /**
+     * Changes the color of the data set.
+     *
+     * @param argb  new color
+     */
+    public void setColor(int argb) {
+        getDataset().setColor(argb);
+    }
+
+    /**
+     * Changes the colors of the Data Series from the passed in Colors List.
+     *
+     * @param colors  List of colors to set to the Data Series
+     */
+    public void setColors(List<Integer> colors) {
+        // With regards to the Colors property setting for the
+        // ScatterChartDataModel, currently an issue exists:
+        // https://github.com/PhilJay/MPAndroidChart/issues/4483
+        // which sets the same color to 2 points at once.
+        getDataset().setColors(colors);
+    }
+
+    /**
+     * Changes the label of the data set.
+     *
+     * @param text  new label text
+     */
+    public void setLabel(String text) {
+        getDataset().setLabel(text);
+    }
+
+    /**
+     * Sets the elements of the Data Series from a CSV-formatted String.
+     *
+     * @param elements String in CSV format
+     */
+    public void setElements(String elements) {
+        // Get the expected number of tuples
+        int tupleSize = getTupleSize();
+
+        // Split all the CSV entries by comma
+        String[] entries = elements.split(",");
+
+        // Iterate over every tuple (by grouping entries)
+        // We start from tupleSize - 1 since the (tupleSize - 1)-th
+        // entry will be the last entry of the tuple.
+        // The index is incremented by the tupleSize to move to the next
+        // group of entries for a tuple.
+        for (int i = tupleSize - 1; i < entries.length; i += tupleSize) {
```

Comment:
@ewpatton Fair point; I guess this was the first thing that came to mind for some reason. I will simplify it to the suggestion.

Reply:
@ewpatton Mainly to handle cases where entries.length is not a multiple of tupleSize.
E.g.: entries.length = 3, tupleSize = 2:
"0, 3, 1"

The fallback option is then to parse (0, 3), and drop 1. The current implementation seemed more simpler to me at that point since the one suggested above would need additional checks on the size of entries.length.

Would you still prefer it to be simplified?